dra-evolution/pkg/api/capacity_types.go

```diff
@@ -3,168 +3,110 @@ package api
+	// - Does it need to be a sub-resource? This would make it harder
+	//   for a driver to publish a new device (spec) and its corresponding
+	//   health information (status).
+}
 
-	// Devices contains the individual devices in the pool.
+type ResourcePoolSpec struct {
+	// NodeName identifies the node which provides the devices. All devices
+	// are local to that node.
 	//
-	// +required
-	// +listType=atomic
-	Devices []Device `json:"devices,omitempty"`
-}
+	// This is currently required, but this might get relaxed in the future.
+	NodeName string `json:"nodeName"`
 
-// DevicePoolStatus contains the state of the pool as last reported by the
-// driver. Note that this will not include the allocations that have been made
-// by the scheduler but not yet seen by the driver. Thus, it is NOT sufficient
-// to make future scheduling decisions.
-type DevicePoolStatus struct {
-	DeviceStatuses []AllocatedDevice `json:"deviceStatuses,omitempty"`
-}
+	// POTENTIAL FUTURE EXTENSION: NodeSelector *v1.NodeSelector
 
-// AllocatedDevice represents a device that has been allocated from the pool.
-type AllocatedDevice struct {
-	Name string
-	// Conditions contains the latest observation of the device's state.
-	Conditions []metav1.Condition `json:"conditions"`
+	// DriverName identifies the DRA driver providing the capacity information.
+	// A field selector can be used to list only ResourcePool
+	// objects with a certain driver name.
+	//
+	// Must be a DNS subdomain and should end with a DNS domain owned by the
+	// vendor of the driver.
+	DriverName string `json:"driverName" protobuf:"bytes,3,name=driverName"`
+
+	// Devices lists all available devices in this pool.
+	Devices []Device `json:"devices,omitempty"`
 
-	// ClaimUIDs contains the UIDs of the claims to which this device
-	// is allocated.
-	ClaimUIDs []types.UID
+	// FUTURE EXTENSION: some other kind of list, should we ever need it.
+	// Old clients seeing an empty Devices field can safely ignore the (to
+	// them) empty pool.
 }
 
-// Device is used to track individual devices in a pool.
+// Device represents one individual hardware instance that can be selected based
+// on its attributes.
 type Device struct {
-	// Name is a driver-specific identifier for the device.
-	// +required
-	Name string `json:"name"`
+	// Name is unique identifier among all devices managed by
+	// the driver on the node. It must be a DNS label.
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
 
-	// Attributes contain additional metadata that can be used in
-	// constraints. If an attribute name overlaps with the pool attribute,
-	// the device attribute takes precedence.
-	//
-	// +optional
-	Attributes []Attribute `json:"attributes,omitempty"`
-
-	// SharedResourcesConsumed contains the pooled resources that are
-	// consumed when this device is allocated.
+	// Attributes defines the attributes of this device.
+	// The name of each attribute must be unique.
 	//
+	// +listType=atomic
```

Comment:
Adding `+listType=map` to get validation seems wrong to me. I still think the main criteria should be "who owns this" and "how is it being updated". Adding `+listType=map` when it is not needed can significantly increase the managed fields annotations because then ownership of each entry gets tracked.

If we ever add automatic validation, then I'd like to see it enabled for a new `+listType=atomicMap` with `+listMapKey=name`.

Reply:
It's bad: 77% of the protobuf encoding is for managed fields. See https://kubernetes.slack.com/archives/C0EG7JC6T/p1717485534890529 in #sig-api-machinery.